crates/bindings-csharp/Runtime/Internal/ITable.cs

```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
 
             public bool MoveNext()
             {
+                if (handle == FFI.RowIter.INVALID)
+                {
+                    return false;
+                }
+
                 uint buffer_len;
                 while (true)
                 {
                     buffer_len = (uint)buffer.Length;
-                    try
+                    var ret = (FFI.Errno)
+                        FFI._row_iter_bsatn_advance(handle, buffer, ref buffer_len);
+                    if (ret == FFI.Errno.EXHAUSTED)
                     {
-                        FFI._iter_advance(handle, buffer, ref buffer_len);
+                        handle = FFI.RowIter.INVALID;
                     }
-                    catch (BufferTooSmallException)
+                    switch (ret)
                     {
-                        buffer = new byte[buffer_len];
-                        continue;
+                        // Iterator exhausted, we're done.
+                        case FFI.Errno.EXHAUSTED:
+                        // Iterator advanced but not exhausted,
+                        // we'll need to advance the iterator in the next call to `MoveNext`.
+                        case FFI.Errno.OK:
+                            // Success! Copy over the row data to `Current` from the scratch `buffer`.
+                            Current = new byte[buffer_len];
+                            Array.Copy(buffer, 0, Current, 0, buffer_len);
+                            return buffer_len != 0;
```

Comment:
FWIW `MoveNext()` should return `false` when the iterator is exhausted. Empty returned buffer used to indicate that before, but IIUC not anymore - now the special status should indicate that instead?

Reply:
Right, but what I meant is what if `ret == 0 && buffer_len == 0`? Right now the iterator will stop on that too - is that something we want intentionally or do we want to stop _only_ on `ret == -1`?

Reply:
I was thinking of an edge case where host - for whatever reason - returned one of the interim buffers as genuinely empty, so `buffer_len == 0` but the iterator is not over yet.

If you think it should be unreachable, let's add `Debug.Assert` to indicate that and to check for that at least in debug builds.

But even with the assertion it still sounds like this particular line can be simplified to `return true;` as the "iterator finished" condition only depends on `handle == FFI.RowIter.INVALID` (after having received "exhausted" return code) and not depend on `buffer_len` at all.

Reply:
> Counter-example: Consider an empty table. You do the equivalent of `MyTable::iter()`. This makes the host side iterator contain zero chunks. A call to `_row_iter_bsatn_advance` will return `-1` the first time and you don't know that `handle == FFI.RowIter.INVALID` yet. Yet, the call to `.MoveNext()` should return `false`, as otherwise, the `foreach` will think there was a chunk/rows, but there wasn't (technically, `while (stream.Position < stream.Length)` is tolerant to this).

You're right, that's actually a great counter-example.

Will `Debug.Assert` suggested above fail on this? Either way, could be nice to have some test for this edge case. I'm guessing you'd like to do that outside this PR, but if so, can you at least create an issue so that we don't forget to add such SDK test later?